### PR TITLE
Lower memory requirement & new optimizer & model save

### PR DESCRIPTION
**Pull request by Siderale**
- Lowered the memory requirement. Instead of 60 images per batch, it is now 40. It now runs on a 1070.
- Fixed some UI glitch with the progress bar.
- Changed the optimizer from SGD to ADAM.
- Added a save method for the best model.
- Accuracy is now the average on the n_episodes.

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -22,19 +22,19 @@
...
-momentum = 0.9
+n_support = 5
+n_query = 5
+n_samples_per_class = n_support + n_query
+learning_rate = 1e-3
 n_epochs = 600
 device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-n_iterations = 100
+n_episodes = 100
+classes_per_it = (40, n_samples_per_class, n_samples_per_class)
```
Comment (FrankZappaLevasseur) [inline src/main.py]: Le nombre de sample par classe c'est pas la même chose que le nombre de classe par itération. Ici, en validation et test, c'est un hasard que les deux valent 10.

File: src/main.py [hunk 2]
```diff
@@ -22,19 +22,19 @@
 valid_path = '../mini_imagenet/csvsplits/valid.csv'
 test_path = '../mini_imagenet/csvsplits/test.csv'
 separator = ';'
-model_best_state_filename = 'best_state.pt'
+filename = 'best_protonet.pt'
```
Comment (FrankZappaLevasseur) [inline src/main.py]: filename c'est plus cute, mais moins explicite (moins claire). Voici ce que Clean Code dit sur les noms de variables : "The name of a variable, function, or class, should answer all the big questions. It should tell you why it exists, what it does, and how it is used."

File: src/main.py [hunk 3]
```diff
@@ -22,19 +22,19 @@
...
 #*############################ 
 #*       Hyperparameters     #
 #*############################
-n_supports = 5
-n_queries = 5
-batch_size = 60
-learning_rate = 0.01
-momentum = 0.9
+n_support = 5
+n_query = 5
```
Comment (FrankZappaLevasseur) [inline src/main.py]: 👍

File: src/main.py [hunk 4]
```diff
@@ -78,13 +75,15 @@
 #*           Entrainement         #
 #*#################################
 history = History()
-best_loss = math.inf
+best_model_weights = None
+best_avg_acc = 0
```
Comment (FrankZappaLevasseur) [inline src/main.py]: 👍

File: src/main.py [hunk 5]
```diff
@@ -99,17 +98,20 @@
...
 
     # if scheduler:
     #       scheduler.step(val_loss)
 
     # Deep copy the best model
-    # if val_loss < best_loss:
-    #     best_loss = val_loss
-    #     best_model_weights = copy.deepcopy(model.state_dict())
-    #     torch.save(best_model_weights, filename)
+    if avg_acc > best_avg_acc:
```
Comment (FrankZappaLevasseur) [inline src/main.py]: Ici, il va falloir vérifier que l'average accuracy calculée sur le jeu de validation (et non training) est supérieur à la meilleur average accuracy, encore une fois calculée sur le jeu de validation.

Mais comme on a dit, on peut rajouter ça dans une autre PR.

Comment (FrankZappaLevasseur) [reply src/main.py]: Pcq comme c'est là, les best weights vont overfitter...

Mais là dessus, je t'apprend rien haha

File: src/main.py (hunk 1 above)
Comment (FrankZappaLevasseur) [reply src/main.py]: Donc en gros, je laisserais 10 hardcodé

File: src/main.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
-# %%
 import math
```
Comment (FrankZappaLevasseur) [inline src/main.py]: probablement à retirer?

File: src/main.py [hunk 7]
```diff
@@ -1,157 +1,45 @@
-import torch
 import copy
-from torch import optim
+import sys
+
+import torch
 from torch.autograd import Variable
 from tqdm import tqdm
```
Comment (FrankZappaLevasseur) [inline src/main.py]: rm les imports copy, sys, variable et tqdm car inutilisés

File: src/utils/meta_train.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import torch
+import copy
+from torch import optim
+from torch.autograd import Variable
```
Comment (FrankZappaLevasseur) [inline src/utils/meta_train.py]: rm les imports torch et optim

File: src/utils/mini_image_net.py [hunk 9]
```diff
@@ -1,18 +1,20 @@
...
 
+from torchvision import transforms
+
 warnings.filterwarnings("ignore")
 
 
 class MiniImageNet(Dataset):
     """Mini Image Net dataset."""
 
-    def __init__(self, csv_file, separator, root_dir, n_supports, n_queries, transforms=None):
```
Comment (FrankZappaLevasseur) [inline src/utils/mini_image_net.py]: :+1:

File: src/utils/mini_image_net.py [hunk 10]
```diff
@@ -22,14 +24,18 @@ def __init__(self, csv_file, separator, root_dir, n_supports, n_queries, transfo
...
-        self.n_queries = n_queries
-        self.transforms = transforms
 
         all_class_paths = [os.path.join(self.root_dir, class_name) for class_name in self.csv_mappings]
         self.all_targets = []
         self.all_image_tensors = []
 
+        load_file_to_pil = lambda filepath: Image.open(filepath)
+        self.transform = transforms.Compose([
+                                        load_file_to_pil,
```
Comment (FrankZappaLevasseur) [inline src/utils/mini_image_net.py]: Bizard de loader l'image dans une transformation... C'est contre-intuitif, mais bon

File: src/utils/few_shot_parameters.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
+import torch
+from torch import optim
+
+from prototypical_loss import PrototypicalLoss
+from utils.dataloader import (load_meta_test_dataloader, load_meta_test_set,
+                              load_meta_train_dataloaders, load_meta_train_set)
+
+
+class FewShotParameters():
```
Comment (FrankZappaLevasseur) [inline src/utils/few_shot_parameters.py]: J'aurais peut-être gardé une seule nomenclature, soit few-shot, soit meta, mais pas les deux... À toi de voir

File: src/utils/few_shot_parameters.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+            self.n_episodes = 100
+            self.PATIENCE_LIMIT = 10
+            self.n_epochs = 10000
+
+            train_set, valid_set = load_meta_train_set(paths)
+
+            sets = {'train_set': train_set,
+                    'valid_set': valid_set
+                   }
```
Comment (FrankZappaLevasseur) [inline src/utils/few_shot_parameters.py]: bizard de closer la bracker comme ca

File: src/utils/few_shot_parameters.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, model, phase, paths):
+        if phase == 'train':
+            learning_rate = 1e-3
+            n_support = 5
+            n_query = 15
+            samples_per_class = n_support + n_query
+            n_ways = (5, 5)
+
+            self.n_episodes = 100
+            self.PATIENCE_LIMIT = 10
```
Comment (FrankZappaLevasseur) [inline src/utils/few_shot_parameters.py]: :+1:

File: src/utils/few_shot_parameters.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class FewShotParameters():
+    def __init__(self, model, phase, paths):
+        if phase == 'train':
+            learning_rate = 1e-3
+            n_support = 5
+            n_query = 15
+            samples_per_class = n_support + n_query
+            n_ways = (5, 5)
```
Comment (FrankZappaLevasseur) [inline src/utils/few_shot_parameters.py]: Donc les ways c'est les classes vu dans une batch? Auquel cas, on passe de (60, 5)  dans le dépôt original à (5, 5) ? 

image
